packages/blunderbuss/src/datastore-lock.ts

```
@@ -0,0 +1,149 @@
+        return true;
+      }
+      if (result === AcquireResult.Failure) {
+        logger.debug(
+          `Failed to acquire the lock, retrying after ${waitTime} miliseconds`
+        );
+      }
+      if (result === AcquireResult.LockActive) {
+        // The lock is active, wait and retry.
+        logger.debug(
+          `The lock is active, retrying after ${waitTime} miliseconds`
+        );
+      }
+      await sleep(waitTime);
+      waitTime = Math.min(
+        maxDelay,
+        waitTime + Math.floor(Math.random() * 2000)
+      );
+    }
+    // Timeout
+    return false;
+  }
+
+  private async _acquire(): Promise<AcquireResult> {
+    const transaction = this.datastore.transaction();
+    try {
+      await transaction.run();
+      // First get the entity
+      const [entity] = await transaction.get(this.key);
+      logger.debug(entity);
+
+      if (entity === undefined || Date.now() - entity.t > this.lockExpiry) {
+        const entity = {
+          key: this.key,
+          data: {
+            t: Date.now(),
+            uuid: this.uniqueId,
+          },
+        };
+        transaction.save(entity);
+        logger.debug(entity);
+        await transaction.commit();
+        return AcquireResult.Success;
+      } else {
+        // The lock is active.
+        await transaction.rollback();
+        return AcquireResult.LockActive;
+      }
+    } catch (err) {
+      err.message = `Error acquiring a lock: ${err.message}`;
+      logger.error(err);
+      await transaction.rollback();
+      return AcquireResult.Failure;
+    }
+  }
+
+  /**
+   * Release the lock.
+   */
+  public async release(): Promise<boolean> {
+    const transaction = this.datastore.transaction();
+    try {
+      await transaction.run();
+      const [entity] = await transaction.get(this.key);
+      if (entity === undefined || entity.uuid !== this.uniqueId) {
+        await transaction.rollback();
+        return true;
+      }
```

This section seems wrong to me - if we attempt to release a lock, it means we think we successfully acquired it earlier. If the lock is acquired by another uuid, doesn't that mean something's gone wrong? Should this throw an error instead?

---

If the lock is acquired by another uuid, this means the lock is expired and another process has obtained the lock. In this case, it should not delete the lock.

---

Isn't this an error condition? It essentially means that blunderbuss didn't acquire the lock for long enough and had exited the critical section prematurely. I think it would be better to throw an error in those situations, so that the error is logged and we can possible address it by increasing the lock timeout

---

Changed to throw an error.